game/game/playercontrol.cpp

```diff
@@ -405,15 +415,21 @@ void PlayerControl::marvinK(uint64_t dt) {
   if (w == nullptr || w->player() == nullptr)
     return;
 
-  auto& pl = *w->player();
+  auto& pl  = *w->player();
   auto  pos = pl.position();
   float rot = pl.rotationRad();
   float s = std::sin(rot), c = std::cos(rot);
-
   Tempest::Vec3 dp(s, 0.0f, -c);
-  pos += dp * 6000 * float(dt) / 1000.f;
+  dp *= 6000 * float(dt) / 1000.f;
 
   pl.clearState(false);
+  if (pl.isSwim()) {
+    auto waterRay = pl.world().physic()->waterRay(pos);
+    auto res      = pl.world().physic()->ray(pos,{pos.x,20000,pos.z});
+    if (!res.hasCol && !waterRay.hasCol)
+      pl.marvinJump();
```

Comment:
Introducing something like `Npc::invalidatePhysics` and removing ray-related logic from PlayerControl feels more natural in such case